File: Assignment_6.R
library(readr)
library(dplyr)
library(tidyr)
library(tidytext)

# Load datasets
speeches <- read_csv("speeches.csv")
fx <- read_csv("fx.csv")

# Keep only columns "date" and "contents" in speeches
speeches <- speeches %>%
  select(date, contents)

# Merge 2 datasets, keeping all the values from fx as the base
merged_data <- left_join(fx, speeches, by = c("DATE" = "date"))

# Searching for outliers - checking basic statistics
summary(merged_data)

# Searching for outliers - checking exchange rate on the boxplot
plot(merged_data$`US dollar/Euro ECB reference exchange rate (EXR.D.USD.EUR.SP00.A)`)

# Update empty records with data from the previous day
clean_data <- merged_data %>%
  arrange(DATE) %>%
  fill(`US dollar/Euro ECB reference exchange rate (EXR.D.USD.EUR.SP00.A)`,
       .direction = "down")

# Checking statistics, including empty cells once again
summary(clean_data)

# Change column name for exchange rate
clean_data <- clean_data %>%
  rename(fx = `US dollar/Euro ECB reference exchange rate (EXR.D.USD.EUR.SP00.A)`)

# Calculating exchange rate return
clean_data <- clean_data %>%
  arrange(DATE) %>%
  mutate(return = (fx - lag(fx)) / lag(fx))

# Extending datasent with 2 new columns concluding good news/bad news on the exchange rate return
clean_data <- clean_data %>%
  mutate(
    good_news = case_when(
      return > 0.005 ~ 1,
      TRUE ~ 0
    ),
    bad_news = case_when(
      return < -0.005 ~ 1,
      TRUE ~ 0
    )
  )

# Remove NA from "contents"
clean_data <- clean_data %>%
  na.omit()

# Creating good_indicators table with the 20 most common words where good_news = 1
good_indicators <- clean_data %>%
  filter(good_news == 1) %>%  #filter on good_news = 1
  unnest_tokens(word, contents) %>% #split text into words
  anti_join(stop_words) %>% #remove stop words
  count(word, sort = TRUE) %>%  #count word repetition
  slice_head(n=20) #select 20 the most occurring words

# Creating bad_indicators table with the 20 most common words where bad_news = 1
bad_indicators <- clean_data %>%
  filter(bad_news == 1) %>% #filter on bad_news = 1
  unnest_tokens(word, contents) %>% #split text into words
  anti_join(stop_words) %>% #remove stop words
  count(word, sort = TRUE) %>% #count word repetition
  slice_head(n=20) #select 20 the most occuring words

# Exporting good_indicators and bad_indicators to CSV
write_csv(good_indicators, "good_indicators.csv")
write_csv(bad_indicators, "bad_indicators.csv")

# Observations: The most common words are very similar for both good and bad news. In both cases the speeches mainly focus on general economic topics such as inflation, policy and growth.
  

File: Assignment_6.py
import pandas as pd
import matplotlib.pyplot as plt #for plotting the graph
import re #for splitting the text into words
from collections import Counter #for counting word repetitions

# Load datasets
speeches = pd.read_csv("speeches.csv")
fx = pd.read_csv("fx.csv")

# Keep only columns "date" and "contents" in speeches
speeches = speeches[["date","contents"]]

# Convert date columns to datetime type
speeches["date"] = pd.to_datetime(speeches["date"])
fx["DATE"] = pd.to_datetime(fx["DATE"])


# Merge 2 datasets, keeping all the values from fx as the base
merged_data = fx.merge(speeches, how="left", left_on="DATE", right_on="date")

# Change column name for exchange rate
merged_data = merged_data.rename(
    columns={"US dollar/Euro ECB reference exchange rate (EXR.D.USD.EUR.SP00.A)": "fx"}
)

# Searching for outliers - checking basic statistics
print(merged_data["fx"].describe())

# Checking no value for the day
merged_data["fx"].isna().sum()

# Searching for outliers - checking exchange rate on a plot
merged_data["fx"].plot()
plt.show()

# Update empty records with data from the previous day
clean_data = merged_data.sort_values("DATE").copy()
clean_data["fx"] = clean_data["fx"].ffill()

# Searching for outliers and no value once again
print(clean_data["fx"].describe())
clean_data["fx"].isna().sum()

# Calculating exchange rate return
clean_data = clean_data.sort_values("DATE").copy()
clean_data["return"] = (clean_data["fx"] - clean_data["fx"].shift(1)) / clean_data["fx"].shift(1)

# Extending dataset with 2 new columns concluding good news/bad news on the exchange rate return
clean_data["good_news"] = (clean_data["return"]>0.005).astype(int)
clean_data["bad_news"] = (clean_data["return"]<-0.005).astype(int)

# Remove NA from contents
clean_data = clean_data.dropna()

# Simple stopwords list
stop_words = {
    "the", "and", "of", "to", "in", "a", "for", "is", "on", "that", "by", "with",
    "as", "are", "at", "an", "be", "this", "from", "or", "it", "we", "our", "their",
    "has", "have", "had", "was", "were", "will", "would", "can", "could", "should",
    "may", "might", "not", "but", "than", "which", "who", "whom", "into", "about",
    "also", "these", "those", "its", "they", "them", "he", "she", "his", "her",
    "you", "your", "i", "my", "me", "us"
}

# Function to get 20 top words
def top_words(df, text_col, top_n = 20):
    text = " ".join(df[text_col].astype(str)).lower()
    words = re.findall(r"\b[a-z]+\b", text)
    words = [word for word in words if word not in stop_words]
    word_counts = Counter(words)
    return pd.DataFrame(word_counts.most_common(top_n), columns=["word", "n"])

# Creating good_indicators table with the 20 most common words where good_news = 1
good_indicators = top_words(clean_data[clean_data["good_news"] == 1], "contents", top_n=20)

# Creating bad_indicators table with the 20 most common words where bad_news = 1
bad_indicators = top_words(clean_data[clean_data["bad_news"] == 1], "contents", top_n=20)

# Exporting good_indicators and bad_indicators to CSV
good_indicators.to_csv("good_indicators.csv", index=False)
bad_indicators.to_csv("bad_indicators.csv", index=False)

# Observations: The most common words are very similar for both good and bad news. In both cases the speeches mainly focus on general economic topics such as inflation, policy and growth.